src/veins/modules/mac/ieee80211p/Mac1609_4.cc

```diff
@@ -567,20 +568,26 @@ void Mac1609_4::handleLowerMsg(cMessage* msg)
 
     EV_TRACE << "Received frame name= " << macPkt->getName() << ", myState= src=" << macPkt->getSrcAddr() << " dst=" << macPkt->getDestAddr() << " myAddr=" << myMacAddr << std::endl;
 
+    bool frameReceived = true;
+    if (dblrand() < frameErrorRate) frameReceived = false;
+
     if (dest == myMacAddr) {
         if (auto* ack = dynamic_cast<Mac80211Ack*>(macPkt)) {
             ASSERT(useAcks);
-            handleAck(ack);
+            if (dblrand() >= ackErrorRate) handleAck(ack);
+            else EV_TRACE << "Artificially dropping ACK";
             delete res;
         }
         else {
-            unique_ptr<BaseFrame1609_4> wsm(check_and_cast<BaseFrame1609_4*>(macPkt->decapsulate()));
-            wsm->setControlInfo(new PhyToMacControlInfo(res));
-            handleUnicast(macPkt->getSrcAddr(), std::move(wsm));
+            if (frameReceived) {
+                unique_ptr<BaseFrame1609_4> wsm(check_and_cast<BaseFrame1609_4*>(macPkt->decapsulate()));
+                wsm->setControlInfo(new PhyToMacControlInfo(res));
+                handleUnicast(macPkt->getSrcAddr(), std::move(wsm));
+            }
```

Comment:
[a] From visual inspection, the else branch might need tidying up of `res` as well. An `EV_TRACE` similar to the case of `ackErrorRate` hitting a frame might be nice, too.